Add DRONE_PROMETHEUS_ANONYMOUS_ACCESS configuration option

JanBerktold opened the pull request.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/drone/drone?pullRequest=2686) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/drone/drone?pullRequest=2686) before we can accept your contribution.<br/><hr/>**Jan Berktold** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/drone/drone?pullRequest=2686) it.</sub>

bradrydzewski commented:
:wave: hey there, have you considered creating a machine account for prometheus access? https://docs.drone.io/administration/user/machine/

bradrydzewski commented:
actually, this url is better https://docs.drone.io/administration/server/metrics/

JanBerktold commented:
Hey @bradrydzewski, we have considered that but we already have Prometheus setup to scrape all Nomad jobs (with a specific tag) for metrics and having to change the central prometheus instance for a single job does not seem ideal and would cause challenges. Our load balancers ensures that no non-internal user can hit the /metrics endpoint to prevent the metrics from leaking to the outside, having unauthenticated access within our system is therefore the preferred option.

bradrydzewski commented on metric/handler.go:
```diff
@@ -26,13 +27,15 @@ var errAccessDenied = errors.New("Access denied")
 
 // NewServer returns a new metrics server.
-func NewServer(session core.Session) *Server {
+func NewServer(session core.Session, config config.Config) *Server {
```
we typically try to avoid passing the entire config into different packages. So in this case I recommend the following for consistency with our other packages:

```
-func NewServer(session core.Session, config config.Config)
+func NewServer(session core.Session, anonymous bool)
```

bradrydzewski replied:
note that this will require modifying https://github.com/drone/drone/blob/master/cmd/drone-server/inject_server.go and creating a provider function. For example:

```diff
// wire set for loading the server.
var serverSet = wire.NewSet(
	manager.New,
-	metric.NewServer,
	api.New,
	web.New,
+	provideMetrics,
	provideRouter,
	provideRPC,
	provideServer,
	provideServerOptions,
)
```

bradrydzewski commented on metric/handler.go:
```diff
@@ -26,13 +27,15 @@ var errAccessDenied = errors.New("Access denied")
 type Server struct {
 	metrics http.Handler
 	session core.Session
+	config  config.Config
```
explained below, we can just store the boolean

```
type Server struct {
  metrics http.Handler
  session core.Session
- config  config.Config
+ anonymous bool
}
```

bradrydzewski commented:
gotcha ... I have no problem with optionally opening up the prometheus endpoint.  I've added some minor comments around the implementation, primarily to ensure consistency with other packages and how we pass around configuration parameters.

JanBerktold commented:
@bradrydzewski Sounds good, working on addressing those right now and also adding a test case. I am unclear on how to inject the bool into metric.NewServer - I haven't worked with wire before but it fails to `inject InitializeApplication: no provider found for bool` which makes sense. Should I just make it an exported field and set it afterwards?

bradrydzewski commented:
you will need to modify this file:
https://github.com/drone/drone/blob/master/cmd/drone-server/inject_server.go

configure to load the metrics server from a custom function:

```diff
// wire set for loading the server.
var serverSet = wire.NewSet(
	manager.New,
-	metric.NewServer,
	api.New,
	web.New,
+	provideMetrics,
	provideRouter,
	provideRPC,
	provideServer,
	provideServerOptions,
)
```

in the same file, implement the custom function:

```
func provideMetrics(session core.Session, config config.Config) *metrics.Server {
  return metrics.New(session, config.Prometheus.AnonymousAccess)
}
```

something like that ... :)

JanBerktold commented:
@bradrydzewski Thanks for the pointers! Sorry for that, haven't worked with wire before and I was a bit blind. :) 

I have addressed the feedback and added a test. It appears that the CI build is failing, but that seems unrelated to my changes. I have confirmed this by pulling the master branch and running the tests locally - they also fail off the unmodified branch.

bradrydzewski commented:
huh strange, last push to master was passing https://cloud.drone.io/drone/drone/127. I will check out the logs and see what I can find.

bradrydzewski commented:
looks maybe like a race condition in one of the unit tests which is why this is intermittent ... I see a nil pointer. I will take care of this and push an update to master, and re-run these tests.

JanBerktold commented:
I spoke too soon - Looks like I had a compilation issue because I didn't regenerate the file. (a git pre-commit hook would be nice).

Unclear on how that could cause a runtime issue though.

bradrydzewski commented:
> Unclear on how that could cause a runtime issue though.

the failed test came from a package untouched by your change, and looks like a race condition, which would explain the intermittent errors. I will handle it separately.

JanBerktold commented:
Looks like we're good here :)

tboerger commented on cmd/drone-server/config/config.go:
```diff
@@ -162,6 +162,11 @@ type (
 
+	// Prometheus provides the prometheus configuration.
+	Prometheus struct {
+		EnableAnonymousAccess bool `envconfig:"DRONE_PROMETHEUS_ANONYMOUS_ACCESS" default:"false"`
```
If you name it AnonymousAcces you don't have to set the envconfig tag.

bradrydzewski commented:
@JanBerktold we received your information request, but the email bounced back. It looks like the mailing list could be restricted. I would be happy to chat, but just need a way to get in touch :)  Perhaps you can shoot a message to brad at drone.io?

badsmoke commented:
why is this not documented?

I have been looking for it for ages
